Concurrent HNSW Merge

zhaih opened the pull request:
### Description

This PR still contains all kinds of magic number and names I need to improve. But I think we can start review it as it is not that hacky and I think is more on acceptable side. For sure I will add quite more javadocs to make the code readable, especially the add graph node part.

The main idea is that we merely need to care about the thread safety of `OnHeapHnswGraph` after the #12651 (still, the entry node election is a bit complex), so I just create one RW lock for each NeighbourArray and lock them when necessary.

The main change happens in `HnswGraphBuilder#addGraphNode` where I changed the sequence of updating the graph such that inserting a new node won't impact searching the previous graph, then I just hardcoded a thread pool in `addVectors` method to test the change. And it shows a good 4x speed up on forceMerge when I'm using 8 threads, and 3x overall indexing speed with 1M docs.

### More details
A new `ConcurrentHnswMerger` to create the `HnswConcurrentMergeBuilder`, which will contain several workers (`ConcurrentMergeWorker`) which essentially each is a graph builder and we will spawn one thread for each of them.

Then main logic still just as stated in the description, I make the `OnHeapHnswGraph` mostly thread-safe, and modified the way we add the graph node from the level 0 and goes up such that when a search is initiated concurrently, this new added node will not lead to a dead-end (because if the new node is discovered, it is guaranteed to have all out-going edges already and it is guaranteed to be able to go to the lower level)

Promoting entry node is a bit tricky because essentially we need to change both node and entry level at the same time. So I made a new class and use AtomicReference to switch it. It seems to be working ok.

One (I think worst) side effect of this change is that we have to manually set the entry node, because we add node to graph from top most level, then add neighbour to the node from the level 0, so we only want to set the new node to be entry node after we have done all this, so it cannot be set automatically when the node is first added at the highest level.

### Benchmark Test

#### Set up
50MB writer buffer, force-merge to 1 segment at last. 384 dimension of vectors.
#### Candidate
```
recall	latency	nDoc	fanout	maxConn	beamWidth	visited	index ms
Force merge done, time: 2 ms
0.838	 0.28	10000	0	64	250	100	3649	1.00	post-filter
Force merge done, time: 20012 ms
0.744	 1.15	100000	0	64	250	100	66680	1.00	post-filter
Force merge done, time: 470161 ms
0.686	12.94	1000000	0	64	250	100	1019300	1.00	post-filter
```

**Update: 100k doc with contention time infomation (note the first few time was for single thread overhead of locks)**
```
Total contention time: 20 ms
Total contention time: 44 ms
Total contention time: 65 ms
Total contention time: 87 ms
Total contention time: 109 ms
Total contention time: 109 ms
Total contention time: 5765 ms
Force merge done, time: 17830 ms
0.741	 1.20	100000	0	64	250	100	65650	1.00	post-filter
```


#### Baseline
```
recall	latency	nDoc	fanout	maxConn	beamWidth	visited	index ms
Force merge done, time: 3 ms
0.838	 0.28	10000	0	64	250	100	3749	1.00	post-filter
Force merge done, time: 97579 ms
0.745	 1.19	100000	0	64	250	100	145355	1.00	post-filter
Force merge done, time: 2059821 ms
0.683	12.81	1000000	0	64	250	100	2902514	1.00	post-filter
```

zhaih committed:
Concurrent HNSW merging, rebased, not tested after rebase

msokolov commented:
This looks like a great start. I worked up a very similar PR but I think I have some concurrency bugs around the update/add of new entry points in the upper graph levels. I might post mine anyway because it has some syntax sugar I think would be nice to bring over here. One idea I had was to create a NeighborArrayIterator that would handle read-locking in the OnHeapGraph case with another implementation that just encapsulates the iteration with no locking for use by other (off-heap) HnswGraph. With that and a little refactoring we can use the same HnswGraphSearcher for both cases?

zhaih commented:
That sounds great, thanks Mike! Please post the PR and I will try to see
how to put it together. Meanwhile let's try to get the prerequisite PR
reviewed and pushed first? Such that this one can look cleaner.

On Thu, Oct 12, 2023, 11:07 Michael Sokolov ***@***.***>
wrote:

> This looks like a great start. I worked up a very similar PR but I think I
> have some concurrency bugs around the update/add of new entry points in the
> upper graph levels. I might post mine anyway because it has some syntax
> sugar I think would be nice to bring over here. One idea I had was to
> create a NeighborArrayIterator that would handle read-locking in the
> OnHeapGraph case with another implementation that just encapsulates the
> iteration with no locking for use by other (off-heap) HnswGraph. With that
> and a little refactoring we can use the same HnswGraphSearcher for both
> cases?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/lucene/pull/12660#issuecomment-1760118913>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEFSB7C4JF6SQ7JQ2SJCR63X7AWVFANCNFSM6AAAAAA55CGFD4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

xjtushilei commented:
Good Job

msokolov commented:
OK, I posted https://github.com/apache/lucene/pull/12683. I'm curious about the performance measurements. We want to know how much lock contention there is so ideally we'd compare total time to merge using single-threaded vs using this. Did your measurement cover the entire indexing process including indexing + flushes (not merges)?

zhaih commented:
Thanks @msokolov, I'll take a look soon.

> so ideally we'd compare total time to merge using single-threaded vs using this.

The most fair comparison here I posted might be the force merge time for 100k docs, which is 20012ms vs 97579ms. Since there should be no merging happening during normal indexing due to index size (around 300mb and per segment 50mb buffer), all the conditions are the same, but still we're counting the flush time and preparation time during the comparison.

Because the lock are only used in several places, I'm wondering whether we could simply just accumulate the lock waitting time using `System.nanoTime()` to get a feel of how much contention I have. But in general a nearly 5x speed up over 8 thread is not bad I feel.

zhaih commented:
@msokolov I incorporate your change about passing in the executor and addVector from range. I also added the wire up to passing in parameters from VectorFormat all the way in.
For the NeighbourIterator I still need to think a bit, the part I'm a little bit unsure is that if we keep the reader lock locked and just pass an iterator out (without copy the array) it may create more contention. So that part I'm still keeping my original implementation.
Anyway I think right now all the prerequisite PRs is checked in so it's in a reviewable state. Altho I still need to add javadocs, tests I still want some early feedback on the class struture, interface and logic improvement. So please review it! 

cc. @benwtrent thanks for creating the merger abstraction, I used it right away, so if you have some spare time please review it as well!

I added some more details to the description you may want to check before reading the code.

benwtrent commented:
This is awesome. I am so happy it's a clean change without tons of complexity and we still get 4x speed up with additional threads.

I will give it a review this weekend or early next week.

msokolov commented on lucene/core/src/test/org/apache/lucene/util/hnsw/HnswGraphTestCase.java:
```diff
@@ -709,6 +710,7 @@ public void testHnswGraphBuilderInvalid() throws IOException {
         IllegalArgumentException.class, () -> HnswGraphBuilder.create(scorerSupplier, 10, 0, 0));
   }
 
+  @Ignore
```
what's the plan for this?

zhaih replied:
Ah, this is something I need some help, I think the real problem here is this line
```
ramUsed(hnsw)
```
Cannot evaluate how much ram is used by the read write lock attached to each neighborArray... So it throws error however I do the estimation...

msokolov replied:
Oh I see, hmm, maybe we need to add Lock to this check in RamUsageTester:

```
    // Ignore JDK objects we can't access or handle properly.
    Predicate<Object> isIgnorable =
        (clazz) -> (clazz instanceof CharsetEncoder) || (clazz instanceof CharsetDecoder);
```

or somehow special case them anyway. Not sure what the usual practice here is; maybe @dweiss can help?

What kind of errors are you seeing?

zhaih replied:
Oh I guess that's the solution, the error I saw was:
```
java.lang.RuntimeException: Can't access field 'readerLock' of class 'java.util.concurrent.locks.ReentrantReadWriteLock' for RAM estimation.
	at __randomizedtesting.SeedInfo.seed([9EFF02730D10D00B:5BBB6633369B07BC]:0)
	at org.apache.lucene.tests.util.RamUsageTester.lambda$createCacheEntry$3(RamUsageTester.java:345)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at org.apache.lucene.tests.util.RamUsageTester.createCacheEntry(RamUsageTester.java:322)
	at org.apache.lucene.tests.util.RamUsageTester.handleOther(RamUsageTester.java:214)
	at org.apache.lucene.tests.util.RamUsageTester.measureObjectSize(RamUsageTester.java:137)
	at org.apache.lucene.tests.util.RamUsageTester.ramUsed(RamUsageTester.java:88)
```
Let me try to add lock to that ignore list

zhaih replied:
Yeah after I added the lock and the AtomicReference to the ignore list it works :)

msokolov reviewed (commented):
mostly minor comments - looks exciting! I would be curious to see the contention times and also understand how this changes CPU usage vs. single-threaded.

zhaih commented on lucene/core/src/java/org/apache/lucene/codecs/lucene95/Lucene95HnswVectorsWriter.java:
```diff
@@ -50,13 +72,24 @@ public final class Lucene95HnswVectorsWriter extends KnnVectorsWriter {
+    } else {
+      exec =
+          Executors.newFixedThreadPool(
+              Runtime.getRuntime().availableProcessors(), new NamedThreadFactory("hnsw-merge"));
```
Yeah you're right.
BTW I'm thinking whether it will be better to have a big threadpool at the Format class and pass it in to every writer?

msokolov replied:
So I *think* that these merge operations are triggered by a single background merge thread? But we shouldn't rely on that and should provide global control over resources dedicate to this, so something like that (thread pool in Format) makes sense. However conceptually the Format doesn't seem right to me - as a user, I would want to be able to control this from IndexWriterConfig along with other merge-related parameters in IndexWriter like buffer size and number of docs before flush. Possibly we could have other concurrent processes running during background merge (like merging different formats in parallel) and we don't want to have separate thread pools for all of those.

zhaih replied:
But this setting should only affect HNSW merge but not necessary all other (potential future) KNN merge? So if we put it to IWC then whether that's a bit too general, so maybe put it into format is ok?

zhaih commented on lucene/core/src/java/org/apache/lucene/codecs/lucene95/Lucene95HnswVectorsWriter.java:
```diff
@@ -635,17 +667,31 @@ private static DocsWithFieldSet writeVectorData(
+      exec.shutdownNow();
+    }
+    System.out.println(
+        "Total contention time: " + NeighborArray.contentionTime.get() / 1000000 + " ms");
```
Let me post one to the description. In general it's normally 5000ms for a 100k docs merge using 8 threads. (single thread time roughly 95000 ms, multithread wall time roughly 17000 ms)

msokolov replied:
Oooh, that's quite a bit! Almost 1/3 of the time spent waiting for locks! I wonder how it would be w/2 threads or 4, and then again with 1M docs. In any case the speedup is great and we can still run the CPU on other tasks while we're waiting.

zhaih replied:
> Almost 1/3 of the time spent waiting for locks!

I don't think so? the 5000 ms is the cpu time and 17k ms is wall time, so roughly if we divide the 5000ms into 8 then it's about 4~5% of total wall time spent in waiting.

msokolov replied:
oh, got it

zhaih commented on lucene/core/src/java/org/apache/lucene/codecs/lucene95/Lucene95HnswVectorsWriter.java:
```diff
@@ -26,17 +26,39 @@
 import org.apache.lucene.codecs.CodecUtil;
 import org.apache.lucene.codecs.KnnFieldVectorsWriter;
 import org.apache.lucene.codecs.KnnVectorsWriter;
-import org.apache.lucene.index.*;
```
LOL I always see some import adjustment after `gradlew tidy`

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswConcurrentMergeBuilder.java:
```diff
@@ -0,0 +1,234 @@
+
+  private final ExecutorService exec;
+  private final ConcurrentMergeWorker[] workers;
+  private InfoStream infoStream = InfoStream.getDefault();
```
It's set by calling `setInfoStream`, see `IncrementalHnswGraphMerger#merge()`

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswConcurrentMergeBuilder.java:
```diff
@@ -0,0 +1,234 @@
+              scorerSupplier.copy(),
+              M,
+              beamWidth,
+              HnswGraphBuilder.randSeed,
```
Yeah I'm not sure what's the best way, but I guess it does not matter that much...

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswConcurrentMergeBuilder.java:
```diff
@@ -0,0 +1,234 @@
+      this.initializedNodes = initializedNodes;
+    }
+
+    private void run(int maxOrd) throws IOException {
```
I think this gives a better load balance, if we divide it upfront, say we give 1/8 number of docs to each thread (like what I did before) then there will likely be some thread finishing way earlier and just wait at the end. Or if we spawn a lot of thread and each give them 2048 docs, I'm worrying it will spend too much memory at last and will make GC unhappy.

This seems to be the best way of doing it IMO, and it's not too complex -- just do a CAS and no more sync is necessary.

And in fact this does give a better performance, previously for 100k docs merge the merge time is around 20k ms, and now it's around 18k ms (in the description see the one with contention time shown)

msokolov replied:
Nice! Let's add a comment here explaining why we did this

zhaih replied:
Sure, let me add some docs around it.

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -33,7 +33,7 @@
  * hyper-parameters.
  */
-public class HnswGraphBuilder {
+public class HnswGraphBuilder implements IHnswGraphBuilder {
```
Yes I noticed that, I use the `IHnswGraphBuilder` just to avoid changing the name of `HnswGraphBuilder`, otherwise it will create a few more diff files that are making the review harder.

I can for sure change the class name accordingly after the code is fully reviewed and approved :)

msokolov replied:
If git is clever it will notice the rename and not show all the lines as changed? Not sure if it is that clever though

zhaih replied:
I'm not sure... But anyway let me keep the name and change it after other parts are reviewed? I'll add a `nocommit`.

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -151,61 +159,124 @@ public OnHeapHnswGraph build(int maxOrd) throws IOException {
+    // hnsw.size());
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
```
It was previous code that's causing the warning. I'll remove it.

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -151,61 +159,124 @@ public OnHeapHnswGraph build(int maxOrd) throws IOException {
-    int[] eps = new int[] {hnsw.entryNode()};
+    // if the entry node is already set, then we have to do all connections first before we can
+    // promote ourselves as entry node
+    // do connections from bottom up
```
Ah this comment was misplaced, sorry for the confusion, I will update it :)

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -151,61 +159,124 @@ public OnHeapHnswGraph build(int maxOrd) throws IOException {
+      for (int level = curMaxLevel; level > nodeLevel; level--) {
+        candidates.clear();
+        graphSearcher.searchLevel(candidates, scorer, level, eps, hnsw, null);
+        eps = new int[] {candidates.popNode()};
```
good catch, let me change it.

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -151,61 +159,124 @@ public OnHeapHnswGraph build(int maxOrd) throws IOException {
+        return;
+      }
+      if (hnsw.numLevels() == curMaxLevel + 1) {
+        throw new IllegalStateException(
```
Yes, if all the logic and calculation is right then this should never happen.

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -221,34 +292,39 @@ private long printGraphBuildStatus(int node, long start, long t) {
         int indexToRemove = findWorstNonDiverse(nbrsOfNbr, nbr);
         nbrsOfNbr.removeIndex(indexToRemove);
       }
+      nbrsOfNbr.rwlock.writeLock().unlock();
```
Yes, thanks for catching that, this is the earlier code that I forgot to improve. Let me change both places

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -221,34 +292,39 @@ private long printGraphBuildStatus(int node, long start, long t) {
-    for (int i = 0; i < size; i++) {
-      int nbr = neighbors.node[i];
+    for (int i = 0; i < scratch.size(); i++) {
+      if (mask[i] == false) {
```
Sure, I added both javadoc and some more comments in `addDiverseNeighbors` method to explain why we need it.

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphMerger.java:
```diff
@@ -0,0 +1,56 @@
+import org.apache.lucene.util.InfoStream;
+
+/**
+ * Abstraction of merging multiple sealed graph into one on heap graph
```
Yeah I'm not sure what's the word here, I guess I just want to say "merge from multiple on-disk graph to one on heap graph"?

msokolov replied:
to be fair, nothing here requires the segments to be merged to be in any particular form as long as they are readable. I would probably just say "merging multiple graphs into one on-heap graph".

zhaih commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswConcurrentMergeBuilder.java:
```diff
@@ -0,0 +1,234 @@
+      }
+      size = neighborArray.size();
+      if (size >= 0) System.arraycopy(neighborArray.node, 0, nodeBuffer, 0, size);
+      neighborArray.rwlock.readLock().unlock();
```
```
Total contention time: 16 ms
Total array copy time: 0 ms
Total contention time: 37 ms
Total array copy time: 0 ms
Total contention time: 50 ms
Total array copy time: 0 ms
Total contention time: 62 ms
Total array copy time: 0 ms
Total contention time: 80 ms
Total array copy time: 0 ms
Total contention time: 80 ms
Total array copy time: 0 ms
Total contention time: 5725 ms
Total array copy time: 1838 ms
Force merge done, time: 18233 ms
0.741	 1.17	100000	0	64	250	100	64325	1.00	post-filter
```

Here's the result.
Still the array copy time is CPU time (accumulated through all threads)

zhaih commented:
> I would be curious to see the contention times and also understand how this changes CPU usage vs. single-threaded.

@msokolov as for CPU usage, I just tested with 1M docs, and on my laptop I can see when it's doing forceMerge it constantly ranges from 790~810%, while if a merge triggers when adding the document, the CPU can goes to 900%.

I would like to see how it scales to, say 20 or even 40 cores, but unfortunately this laptop with 10 cores (including 2 efficient core) is the only one I have for now...

BTW the forceMerge of the 1M index took `362125 ms`, while the contention time for that merge was `81847 ms` in CPU time, so average `10230 ms` wall time and thus `~2.8%` of total time. And array copy time was `12151 ms` in CPU time, and average `1519 ms` wall time and thus `~0.4%` of total time. So you're right, comparing with result of 100k documents, the contention overhead goes down when the graph is getting bigger.

benwtrent commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -221,34 +296,50 @@ private long printGraphBuildStatus(int node, long start, long t) {
-      if (nbrsOfNbr.size() > maxConnOnLevel) {
-        int indexToRemove = findWorstNonDiverse(nbrsOfNbr, nbr);
-        nbrsOfNbr.removeIndex(indexToRemove);
+      long start = System.nanoTime();
+      nbrsOfNbr.rwlock.writeLock().lock();
```
My small concern here is that now we are locking even when no executor is being used and we know its a single thread. How bad are we hitting users that aren't using multi-threaded merge?

zhaih replied:
So when I'm testing it's like 20ms of overhead on grabing this lock for building roughly 20k docs with 384 dim vectors. So I think it's not bad given the HNSW build itself is much slower?

benwtrent replied:
>  20ms of overhead on grabing this lock for building roughly 20k docs with 384 dim vectors. 

Does contention grow linearly? Or is it logarithmic?

msokolov replied:
I had previously posted a NeighborIterator abstraction that was designed to hide the locking (somewhat) from the HnswSearcher/Builder. That way we don't do any locking in the serial implementation. But @zhaih pointed out it was too aggressively holding the lock around a large critical section. I think we could enhance the concurrent implementation of that abstraction by having it copy the neighbors as this PR does, and then release the lock immediately after copying rather than holding the lock for the entire iteration.

zhaih replied:
> Does contention grow linearly? Or is it logarithmic?

@benwtrent If you mean the 20ms it's growing linearly with number of nodes we insert, as it's basically the overhead we pay for acquiring the lock for each neighbor. It's not contention tho because it's single thread execution.

zhaih replied:
@msokolov Yeah that solves the searcher's part, however I think the more complex one is on writer as I mentioned below we need to group multiple method together in a "transaction". 
I think in you PR you used a returned Closeable to kinda "hide" the lock, but I think that's essentially the same as exposing the lock?

benwtrent commented on lucene/core/src/java/org/apache/lucene/util/hnsw/NeighborArray.java:
```diff
@@ -35,6 +38,9 @@ public class NeighborArray {
   float[] score;
   int[] node;
   private int sortedNodeSize;
+  public final ReadWriteLock rwlock = new ReentrantReadWriteLock(true);
```
Why don't we have a `ConcurrentNeighborArray`? Always requiring locking by the external caller is not good. NeighborArray needs to change its API so external folks shouldn't worry about the lock.

zhaih replied:
I think we can hide the lock inside the `NeighborArray` by adding some APIs, but it's hard to make a totally thread safe class given how we use it right now: for example, when we add a neighbor and need to purge the non-diverse one, we're adding -> sorting -> calculating non-diverse -> removing. And it's not enough to just ensure each operation to be atomic because someone can always add a node after we sort and makes things more complex. So even I create this `ConcurrentNeighborArray` and every operation is nicely protected the caller still need to be extra careful...

But with the caller control the lock, it's much easier to control this "transaction" behavior and only add lock when they need.

I'm not saying this is good, but given this is really a very hacky class already (even before the lock) and internal to HNSW only maybe we can leave it as a follow-up?

benwtrent replied:
We can make it a refactor later, but I am convinced we will want a separate `*NeighborArray` class for concurrent vs. serial graph building. There is no technical reason to slow down serial graph building and I think we can make concurrent even nicer by freeing itself from having to even think about serial graph building.

zhaih replied:
Yeah I don't have a very good idea for that yet, I'll for sure create an issue after this is merged.

benwtrent replied:
It seems like we need a `ConcurrentHNSWGraphBuilder`. It can use much of the same logic, but it instead uses concurrent neighbors. But that can be a later refactor.

benwtrent commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphMerger.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @lucene.experimental
+ */
+public interface HnswGraphMerger {
```
NICE!

benwtrent reviewed (commented):
Comments on API design. I am not sure if this is considered a "final draft for review" or whats just there as a stop-gap for testing still.

zhaih commented on lucene/core/src/java/org/apache/lucene/codecs/lucene95/Lucene95HnswVectorsFormat.java:
```diff
@@ -146,18 +148,24 @@ public final class Lucene95HnswVectorsFormat extends KnnVectorsFormat {
    * @param beamWidth the size of the queue maintained during graph construction.
    */
-  public Lucene95HnswVectorsFormat(int maxConn, int beamWidth) {
+  public Lucene95HnswVectorsFormat(int maxConn, int beamWidth, int mergeWorkers) {
```
Yes Mike is also having the same comment, but I'm not very sure we want the setting to be in IWC because this is specifically for HNSW, and in future if we are going to have more KNN implmentation then this setting is not general enough to be in IWC?

benwtrent replied:
@zhaih I understand. We don't need to make the bigger move until we see the need for parallel merges elsewhere.

But, I do think it should be an `ExecutorService` that is passed in here, not a number of workers.

msokolov replied:
eventually I would think we could have something like `IndexWriterConfig.setMergeThreadConcurrency(int)` that would provide a shared thread pool for any work done during merging. Or maybe it could accept an ExecutorService. But agreed since we only have this one use now we could tie it to the KnnVector/Hnsw API. A consequence of exposing this on the format is you would be able to turn it on and off per-field. Is that a benefit or a bug?

zhaih replied:
I wired an `ExecutorService` from format :)

zhaih commented:
@msokolov @benwtrent I removed almost all `nocommit` (except the renaming one) and rebased to main, please take a look if you have time.

@benwtrent Please check whether the rebase and modification, especially the scalar quantized part is done correctly. I did it according to my understanding but that is not necessarily correct :)

I'll do another round of benchmark run tomorrow to make sure things are still intact.

zhaih committed:
fix small bug

benwtrent commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -151,61 +159,128 @@ public OnHeapHnswGraph build(int maxOrd) throws IOException {
+    if (infoStream.isEnabled(HNSW_COMPONENT)) {
+      infoStream.message(HNSW_COMPONENT, "addVectors [" + minOrd + " " + maxOrd + ")");
+    }
+    //    System.out.println("addVectors [" + minOrd + " " + maxOrd + ") initialized.size=" +
+    // initializedNodes.size());
```
all these comments should probably be removed.

zhaih replied:
Good catch, removed!

benwtrent commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsFormat.java:
```diff
@@ -198,14 +218,25 @@ public Lucene99HnswVectorsFormat(
   @Override
   public KnnVectorsWriter fieldsWriter(SegmentWriteState state) throws IOException {
-    return new Lucene99HnswVectorsWriter(state, maxConn, beamWidth, scalarQuantizedVectorsFormat);
+    return new Lucene99HnswVectorsWriter(
```
seems correct to me!

zhaih replied:
Thank you!

benwtrent commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsWriter.java:
```diff
@@ -557,6 +566,12 @@ public void close() throws IOException {
                 segmentWriteState.directory.deleteFile(tempFileName);
               }
+
+              @Override
```
yeah, this seems ok. We just need to be able to close the file handles and delete the temporary file. The internal copies should only be able to read and score while these still exist. 

If close occurs and something is still hanging onto a copy, it should be undefined and throw. Especially since these are internal/experimental objects, this is OK by me.

zhaih committed:
remove extra comments, tested

zhaih commented:
I reran the benchmark with the current code and it still looks good, both the recall and latency remains the same.

The only unresolved issue right now is the naming of the interface, I'm aware that `IHnswGraphBuilder` is not the convention in lucene, however if we change the `HnswGraphBuilder` to some other name and change the interface to `HnswGraphBuilder` git will not be smart enough to know the renaming and migrate the history. So any idea on what will be a good name of the intterface?

zhaih committed:
Fix back comp

zhaih committed:
rename

benwtrent reviewed (approved):
@zhaih could you create an issue for tracking us optimizing the single threaded case to not use locks?

msokolov commented:
as for the renaming maybe we can look to do it as a followup PR? IHnswGraphSearcher -> HnswGraphSearcher? Or maybe we can just use HnswSearcher right now?

Re: controlling concurrency. Maybe eventually we figure out how to do this as part of ConcurrentMergeScheduler? Since it is all about merging, that seems like a logical place. It currently allows setting `maxMergeCount` and `maxThreadCount` independently, but enforces `maxThreadCount <= maxMergeCount`. We could potentially fiddle with this, maybe it would be as simple as relaxing that constraint and using "spare threads" to do concurrent graph merging.

msokolov reviewed (approved):
Looking good, I think it's ready to merge.

zhaih committed:
Rename

zhaih commented:
Thanks Mike and Ben for reviewing! I'll try to merge and backport it tomorrow. And create a issue for future refactoring

zhaih committed:
javadoc

zhaih committed:
CHANGES entry

zhaih commented:
FYI I pushed a commit https://github.com/apache/lucene/commit/4576ae09e8885f40cc27424fa8d529aa5c172422 to fix the behavior that vector writer close the executor passed in

zhaih commented:
> For my understanding, how does it play out with merge throttling, could we end up sleeping within tasks that get passed to the configured executor service?

I think we rate limit the IndexOutput only in CMS? If my understand is correct then **I think** this is not affected by throttling as it's all memory operation and we'll write to disk after all the tasks are completed
